src/txmempool.cpp

```diff
@@ -99,96 +139,338 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
+    const CScCertificate& cert = mapCertificate[hash].GetCertificate();
+
+    for (unsigned int i = 0; i < cert.GetVin().size(); i++)
+        mapNextTx[cert.GetVin()[i].prevout] = CInPoint(&cert, i);
 
-void CTxMemPool::remove(const CTransaction &origTx, std::list<CTransaction>& removed, bool fRecursive)
+    if (mapSidechains.count(cert.GetScId()) )
+        assert(mapSidechains[cert.GetScId()].backwardCertificate.IsNull());
+    else
+        LogPrint("cert", "%s():%d - adding [%s] in mapSidechain\n", __func__, __LINE__, cert.GetScId().ToString() );
+
+    mapSidechains[cert.GetScId()].backwardCertificate = hash;
+           
+    nCertificatesUpdated++;
+    totalCertificateSize += entry.GetCertificateSize();
+    cachedInnerUsage += entry.DynamicMemoryUsage();
+    // TODO cert: for the time being skip the part on policy estimator, certificates currently have maximum priority
+    // minerPolicyEstimator->processTransaction(entry, fCurrentEstimate);
+    LogPrint("cert", "%s():%d - cert [%s] added in mempool\n", __func__, __LINE__, hash.ToString() );
+    return true;
+}
+
+void CTxMemPool::remove(const CTransactionBase& origTx, std::list<CTransaction>& removedTxs, std::list<CScCertificate>& removedCerts, bool fRecursive, bool removeDependantFwds)
 {
     // Remove transaction from memory pool
     {
         LOCK(cs);
-        std::deque<uint256> txToRemove;
-        txToRemove.push_back(origTx.GetHash());
-        if (fRecursive && !mapTx.count(origTx.GetHash())) {
+        std::deque<uint256> objToRemove;
+        objToRemove.push_back(origTx.GetHash());
+
+        if(!fRecursive)
+            return removeInternal(objToRemove, removedTxs, removedCerts, fRecursive, removeDependantFwds);
+
+        if (!mapCertificate.count(origTx.GetHash()) && !mapTx.count(origTx.GetHash())) {
             // If recursively removing but origTx isn't in the mempool
             // be sure to remove any children that are in the pool. This can
-            // happen during chain re-orgs if origTx isn't re-accepted into
+            // happen during chain re-orgs if origCert isn't re-accepted into
             // the mempool for any reason.
-            for (unsigned int i = 0; i < origTx.vout.size(); i++) {
+            for (unsigned int i = 0; i < origTx.GetVout().size(); i++) {
                 std::map<COutPoint, CInPoint>::iterator it = mapNextTx.find(COutPoint(origTx.GetHash(), i));
                 if (it == mapNextTx.end())
                     continue;
-                txToRemove.push_back(it->second.ptx->GetHash());
+                LogPrint("sc", "%s():%d - adding tx [%s] to list for removing\n", __func__, __LINE__, it->second.ptx->GetHash().ToString());
+                objToRemove.push_back(it->second.ptx->GetHash());
+            }
+
+            for(const auto& sc: origTx.GetVscCcOut()) {
+                if (mapSidechains.count(sc.scId) == 0)
+                    continue;
+                if (removeDependantFwds) {
```

Comment:
instead of if else just directly schedule removal of all txs that where sending funds to this sc
 for(const auto& fwdTxHash : mapSidechains.at(sc.scId).fwdTransfersSet)
                        objToRemove.push_back(fwdTxHash);

Reply:
Done